qcodes/instrument/base.py

```diff
@@ -434,7 +433,36 @@ def validate_status(self, verbose: bool = False) -> None:
                 p.validate(value)
 
 
-class AbstractInstrument(ABC):
+class AbstractInstrumentMeta(ABCMeta):
+    """
+    Metaclass used to customize Instrument creation. We want to register the
+    instance iff __init__ successfully runs, however we can only do this if
+    we customize the instance initialization process, otherwise there is no
+    way to run `register_instance` after `__init__` but before the created
+    instance is returned to the caller.
+
+    Instead we use the fact that `__new__` and `__init__` are called inside
+    `type.__call__`
+    (https://github.com/python/cpython/blob/main/Objects/typeobject.c#L1077)
```

Comment:
could you add a link to https://github.com/python/typeshed/blob/master/stdlib/builtins.pyi#L156
too

That shows how mypy expects call to be typed. Which is consistent with this